mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/auth/firebase/FirebaseAuth.kt

```diff
@@ -31,35 +31,59 @@ class FirebaseAuth(private val actual: ActualFirebaseAuth) : Auth {
             .buffer(capacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
             .map { it?.let(::FirebaseUser) }
 
+  /** An enumeration representing the [AuthenticationMode] that a user chose. */
+  private enum class AuthenticationMode {
+
+    /** The mode in which a user wants to login to their existing account. */
+    SignIn,
+
+    /** The mode in which a user wants to create a new account. */
+    SignUp,
+  }
+
   /**
    * Runs the [block] and maps the resulting [AuthResult] to an [AuthenticationResult].
    *
+   * @param mode the [AuthenticationMode] used to authenticate.
    * @param block the block of code which is used to interact with Firebase.
    */
   private suspend fun authenticate(
+      mode: AuthenticationMode,
       block: suspend () -> AuthResult?,
   ): AuthenticationResult =
       try {
         val result = block()
         val user = result?.user?.let(::FirebaseUser)
         AuthenticationResult.Success(user)
       } catch (other: Throwable) {
-        AuthenticationResult.FailureInternal
+        when (other) {
+          is FirebaseAuthWeakPasswordException -> AuthenticationResult.Failure.BadPassword
+          is FirebaseAuthInvalidCredentialsException ->
+              when (mode) {
+                AuthenticationMode.SignIn -> AuthenticationResult.Failure.IncorrectPassword
+                AuthenticationMode.SignUp -> AuthenticationResult.Failure.IncorrectEmailFormat
+              }
+          is FirebaseAuthUserCollisionException -> AuthenticationResult.Failure.ExistingAccount
+          is FirebaseAuthInvalidUserException -> AuthenticationResult.Failure.InvalidUser
+          else -> AuthenticationResult.Failure.Internal
```

Comment:
Nitpick : Import `AuthenticationResult.Failure.*` to have something more readable